File: boj/graph/16940.java
import java.util.*;

public class Main {

    static int N;
    static ArrayList<Integer>[] rel;
    static int[] result;
    static int[] parent;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        N = sc.nextInt();
        rel = new ArrayList[N + 1];
        result = new int[N + 1];
        parent = new int[N + 1];

        for (int i = 1; i <= N; i++) {
            rel[i] = new ArrayList<>();
        }

        for (int i = 0; i < N - 1; i++) {
            int e1 = sc.nextInt();
            int e2 = sc.nextInt();

            rel[e1].add(e2);
            rel[e2].add(e1);
        }

        for (int i = 1; i <= N; i++) {
            result[i] = Integer.parseInt(sc.next());
        }

        System.out.printf("%d", search() ? 1 : 0);
    }

    static boolean search() {
        Queue<Integer> que = new LinkedList<>();
        boolean[] visited = new boolean[N + 1];

        que.add(1);
        visited[1] = true;

        int index = 2;
        for (int i = 1; i <= N; i++) {
            int cur = que.poll();

            int cnt = 0; // 자식 노드의 개수
            for (int child : rel[cur]) {
                if (visited[child]) continue;
                cnt++;
                parent[child] = cur;
            }

            for (int j = index; j < index + cnt; j++) {
                // 연결 안되거나 범위를 벗어날 때는 return false;
                if (j > N || parent[result[j]] != cur) return false;
                que.add(result[j]);
                visited[result[j]] = true;
            }

            index += cnt;
        }

        return true;
    }

}

File: boj/graph/16940.swift
import Foundation

guard let inputOfN = readLine() else { exit(0) }
guard let N = Int(inputOfN) else { exit(0) }

var map = [[Int]](repeating: [Int](), count: N+1)

for _ in 0..<N-1 {
    guard let input = readLine()?.split(separator: " ").compactMap({Int(String($0))}) else { exit(0) }
    
    map[input[0]].append(input[1])
    map[input[1]].append(input[0])
}

guard let target = readLine()?.split(separator: " ").compactMap({Int(String($0))}) else { exit(0) }

print(search() ? "1" : "0")

func search() -> Bool {
    
    var que = [Int]()
    var visited = [Bool](repeating: false, count: N+1)
    var parent = [Int](repeating: 0, count: N+1)
    
    que.append(1)
    visited[1] = true
    
    var index = 2
    for _ in 0..<N {
        let cur = que.removeFirst()
        
        var count = 0
        for element in map[cur] {
            if visited[element] { continue }
            parent[element] = cur
            count+=1
        }
        
        for i in index..<index+count {
            if i >= N+1 || parent[target[i-1]] != cur { return false }
            que.append(target[i-1])
            visited[target[i-1]] = true
        }
        
        index+=count
    }
    
    return true
}
